state/relationunit.go

```diff
@@ -652,36 +652,28 @@ func NetworksForRelation(
 			if err != nil {
 				return "", nil, nil, errors.Trace(err)
 			}
-
 			machine, err := st.Machine(machineID)
 			if err != nil {
 				return "", nil, nil, errors.Trace(err)
 			}
-
 			networkInfos := machine.GetNetworkInfoForSpaces(set.NewStrings(boundSpace))
 			// The binding address information based on link layer devices.
 			for _, nwInfo := range networkInfos[boundSpace].NetworkInfos {
 				for _, addr := range nwInfo.Addresses {
 					ingress = append(ingress, addr.Address)
 				}
-
-			}
-		} else {
-			// Be be consistent with IAAS behaviour above, we'll return all
-			// addresses, including any container address.
-			addr, err := unit.AllAddresses()
-			if err != nil {
-				logger.Warningf(
-					"no service address for unit %q in relation %q",
-					unit.Name(), rel)
-			} else {
-				network.SortAddresses(addr)
-				for _, a := range addr {
-					ingress = append(ingress, a.Value)
-				}
 			}
 		}
 	}
+	if len(ingress) == 0 {
+		address, err := fetchAddr(unit.PrivateAddress)
```

Comment:
This also should be up in the if crossmodel{} block to replace address, err = unit.PrivateAddress(). We don't want to delete that logic.
We should retain the if unit.ShouldBeAssigned() {} else {} logic and have this in the else{} bit.